frontend/middle-3/fp.md

```diff
@@ -0,0 +1,183 @@
+# Functional programming
+
+## Closures
+* What are free variables?
+* How can immutable closures be useful? Give several examples.
+* Dynamic scoping
+  * What is dynamic scoping?
+  * Why is it inconvenient?
+  * What are currently widely used analogues?
+* Does JS have dynamic or lexical scoping? Is there some exceptions (consider `this` for example)?it
+* What is the funarg problem?
+  * Why is it divided to upward and downward problems?
+  * Is funarg problem possible in languages without support of free variables?
+  * Is funarg problem possible in languages where function cannot escape its scope?
+  * Is funarg problem possible in languages with dynamic scoping?
+* How closures can be emulated? What are problems will arise with those approaches?
+* Why does mutable closure is a sufficient tool to add concurrency to the language (give examples with JS)?
+* Compare closures and objects? What are advantages of each one?
+
+### Resources
+* [JS scope: static, dynamic, and runtime-augmented](https://codeburst.io/js-scope-static-dynamic-and-runtime-augmented-5abfee6223fe)
+* [Notes from the JS pit: closure optimization](http://blog.cdleary.com/2010/05/notes-from-the-js-pit-closure-optimization/)
+
+
+## Recursion
+TODO — fill the recursion questions
+* What are "structural" and "generative" types of recursion?
+
+## Tail call
+* What is the tail call optimization? Why is it called "tail"? What is the problem with recursions without TCO?
+* Is call stack still needed for the languages with TCO? What's a problem with existence of call stacks in such languages?
+* What is the difference between tail recursion elimination and tail call optimization?
+* Are tail calls only about looping? Can tail calls provide us with something more powerfull?
```

Comment:
Can tail calls be only used for loops? Can they provide us with something more powerful?